### PR TITLE
Lotus architecture notes

### PR DESCRIPTION
**Pull request by schomatis**
Read rendered version [here](https://github.com/filecoin-project/lotus/blob/schomatis/feat/docs/dev/documentation/en/dev/architecture-2.md).

--------------------------------------------------------------------------

Just some loose notes, not a coherent narrative yet.

Moved some doc files around which might have broken a few links.

Towards #1547.

### CODE REVIEW

**Comment by schomatis** [comment]
@arajasek Feel free to modify anything and push directly to the branch, I'll add new content through PR comments but will be mostly independent of what's already there so do as you think is best.

File: documentation/en/dev/architecture.md [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+
+The first abstraction is the `Blockstore` interface (`github.com/ipfs/go-ipfs-blockstore/blockstore.go`) which structures the key-value pair into the CID format for the key and the `Block` interface (`github.com/ipfs/go-block-format/blocks.go`) for the value, which is just a raw string of bytes addressed by its hash (which is included in the CID key/identifier). `ChainBlockstore` will create a `Blockstore` in the repo under the `/blocks` namespace (basically every key stored there will have that prefix so it does not collide with other stores that use the same underlying repo).
+
+FIXME: Link to IPFS documentation about DAG, CID, and related, especially we need a diagram that shows how do we wrap each datastore inside the next layer (datastore, batching, block store, gc, etc).
+
+Similarly, `modules.Datastore()` creates a `dtypes.MetadataDS` (alias for the basic `Datastore` interface) to store metadata under the `/metadata`. (FIXME: Explain *what* is metadata in contrast with the block store, namely we store the pointer to the heaviest chain, we might just link to that unwritten section here later.)
+
+FIXME: Explain the key store related calls.
```
Comment (schomatis) [inline documentation/en/dev/architecture.md:101]: We can probably remove this, will describe the key store in a separate document explaining how to create a miner node.

File: documentation/en/dev/architecture.md [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+
+It is important to note at this point that similar to the IPFS architecture of addressing by content and not by location/address (FIXME: check and link to IPFS docs) the "actual" chain stored in the node repo is *relative* to which CID we look for. We always have stored a series of Filecoin blocks pointing to other blocks, each a potential chain in itself by following its parent's reference, and its parent's parent, and so on up to the genesis block. (FIXME: We need a diagram here, one of the Filecoin blog entries might have something similar to what we are describing here.) It only depends on *where* (location) do we start to look for. The *only* address/location reference we hold of the chain, a relative reference, is the `heaviest` pointer. This is reflected by the fact that we don't store it in the `Blockstore` by a fixed, *absolute*, CID that reflects its contents, as this will change each time we sync to a new head (FIXME: link to the immutability IPFS doc that I need to write).
```
Comment (schomatis) [inline documentation/en/dev/architecture.md:235]: Mutation of state in IPFS: https://github.com/ipfs/ipfs-docs/issues/264

(For now let's just point to the issue with my notes until someone from IPFS docs can polish it and merge it into the site.)

File: documentation/en/dev/architecture-2.md [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+Note: The API refers to this stage as `StageHeaders`.
+
+We proceed in the sync process by requesting block headers from the peer, 
+moving back from their head, until we reach a tipset that we have in common
+(such a common tipset must exist, thought it may simply be the genesis block).
+
+Some of the possible causes of failure in this stage include:
+
+- The chain is linked to a block that we have previously marked as bad, and stored in a `BadBlockCache`.
+- The beacon entries in a block are inconsistent (WHYMAGIK: more details about what is validated here wouldn't be bad).
```
[the remark this answers is not in the document]
Comment (arajasek) [reply documentation/en/dev/architecture-2.md]: I'd like to leave something about "how can this go wrong", unless it'll be superseded by the validation document. Gonna leave this as is for now, with a note slapped on top saying we might want to replace the para with a link to the validation doc.

**Comment by arajasek** [comment]
All right:

- Addressed all the review comments
- Reorganised stuff so that only the fresh documentation will go live on lotu.sh
- Cleaned up the `dev` folder to keep some of the content we might want later

**Comment by schomatis** [comment]
Ok, if the `dev/` is mostly a scratchpad directory for now I'll revert the renames (to hopefully not break anything else) and merge.

**Pull request merged**